File: src/backtracking/cpp/minimax/test/test_minimax.cpp
#include "minimax.h"
#include "gtest/gtest.h"

TEST(TestMinimax, TestEmptyBoard) {
  std::vector<std::vector<int>> board = {
      {0, 0, 0},
      {0, 0, 0},
      {0, 0, 0},
  };
  int player = -1;
  std::pair<int, int> expected = {0, 0};
  std::pair<int, int> actual = find_best_move(board, player);

  EXPECT_EQ(expected, actual);
}

TEST(TestMinimax, TestTwoMovesAlreadyMade) {
  std::vector<std::vector<int>> board = {{-1, 0, 0}, {0, 1, 0}, {0, 0, 0}};
  int player = -1;

  std::pair<int, int> expected = {0, 1};
  std::pair<int, int> actual = find_best_move(board, player);

  EXPECT_EQ(expected, actual);
}

TEST(TestMinimax, TestDraw) {
  std::vector<std::vector<int>> board = {{-1, 1, -1}, {1, -1, 1}, {-1, 1, -1}};
  int player = -1;

  std::pair<int, int> expected = {-1, -1};
  std::pair<int, int> actual = find_best_move(board, player);

  EXPECT_EQ(expected, actual);
}

TEST(TestMinimax, TestPlayerAWins) {
  std::vector<std::vector<int>> board = {{-1, 1, -1}, {1, 1, 0}, {-1, 0, -1}};
  int player = -1;
  std::pair<int, int> expected = {1, 2};
  std::pair<int, int> actual = find_best_move(board, player);

  EXPECT_EQ(expected, actual);
}

File: src/backtracking/python/minimax/tests/test_minimax.py
import unittest
from src.minimax import minimax, find_best_move


class TestMinimax(unittest.TestCase):
    def test_empty_board(self):
        board = [
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
        ]
        player = -1
        expected = (0, 0)
        actual = find_best_move(board, player)

        self.assertEqual(expected, actual)

    def test_two_moves_already_made(self):
        board = [[-1, 0, 0], [0, 1, 0], [0, 0, 0]]
        player = -1

        expected = (0, 1)
        actual = find_best_move(board, player)

        self.assertEqual(expected, actual)

    def test_draw(self):
        board = [[-1, 1, -1], [1, -1, 1], [-1, 1, -1]]
        player = -1

        actual = find_best_move(board, player)

        self.assertIsNone(actual)

    def test_player_a_wins(self):
        board = [[-1, 1, -1], [1, 1, 0], [-1, 0, -1]]
        player = -1
        expected = (1, 2)
        actual = find_best_move(board, player)

        self.assertEqual(expected, actual)
